compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3403,146 +3403,441 @@ TR::Register *OMR::X86::TreeEvaluator::encodeUTF16Evaluator(TR::Node *node, TR::
+   switch (valueReg->getKind())
+      {
+      case TR_GPR:
+         generateRegRegInstruction(MOVDRegReg4, node, XMM, valueReg->getHighOrder(), cg);
+         generateRegImmInstruction(PSLLQRegImm1, node, XMM, 32, cg);
+         generateRegRegInstruction(MOVDRegReg4, node, XMM, valueReg->getLowOrder(), cg);
+         break;
+      case TR_FPR:
+         generateRegRegInstruction(MOVDQURegReg, node, XMM, valueReg, cg);
+         break;
+      default:
+         TR_ASSERT(0, "Arrayset Evaluator: unsupported register type");
+         break;
+      }
+
+   // Store the XMM register to memory via a loop
+   // Example:
+   //   SHR RCX,3
+   //   JRCXZ endLabel
+   // loopLabel:
+   //   MOVQ [RDI+8*RCX-8],XMM0
+   //   LOOP loopLabel
+   // endLable:
+   //   # LOOP END
+   generateRegImmInstruction(SHRRegImm1(), node, sizeReg, 3, cg);
+   generateLabelInstruction(JRCXZ1, node, endLabel, cg);
+   generateLabelInstruction(LABEL, node, loopLabel, cg);
+   generateMemRegInstruction(MOVQMemReg, node, generateX86MemoryReference(addressReg, sizeReg, 3, -8, cg), XMM, cg);
+   generateLabelInstruction(LOOP1, node, loopLabel, cg);
+
+   generateLabelInstruction(LABEL, node, endLabel, deps, cg);
+   cg->stopUsingRegister(XMM);
+   }
+
+inline void ArraySetDefault(TR::Node* node, uint8_t elementSize, TR::Register* addressReg, TR::Register* valueReg, TR::Register* sizeReg, TR::CodeGenerator* cg)
+   {
+   TR::Register* EAX = cg->allocateRegister();
+
+   TR::RegisterDependencyConditions* stosDependencies = generateRegisterDependencyConditions((uint8_t)0, (uint8_t)3, cg);
+   stosDependencies->addPostCondition(EAX, TR::RealRegister::eax, cg);
+   stosDependencies->addPostCondition(sizeReg, TR::RealRegister::ecx, cg);
+   stosDependencies->addPostCondition(addressReg, TR::RealRegister::edi, cg);
+
+   // Load value to EAX
+   switch (valueReg->getKind())
+      {
+      case TR_GPR:
+         generateRegRegInstruction(MOVRegReg(elementSize == 8), node, EAX, valueReg, cg);
+         break;
+      case TR_FPR:
+         generateRegRegInstruction(elementSize == 8 ? MOVQReg8Reg : MOVDReg4Reg, node, EAX, valueReg, cg);
+         break;
+      default:
+         TR_ASSERT(0, "Arrayset Evaluator: unsupported register type");
+         break;
+      }
+
+   // Store EAX into memory
+   switch (elementSize)
       {
       case 1:
+         generateInstruction(REPSTOSB, node, stosDependencies, cg);
          break;
       case 2:
          generateRegImmInstruction(SHRRegImm1(cg), node, sizeReg, 1, cg);
+         generateInstruction(REPSTOSW, node, stosDependencies, cg);
          break;
       case 4:
          generateRegImmInstruction(SHRRegImm1(cg), node, sizeReg, 2, cg);
+         generateInstruction(REPSTOSD, node, stosDependencies, cg);
          break;
       case 8:
          generateRegImmInstruction(SHRRegImm1(cg), node, sizeReg, 3, cg);
+         generateInstruction(REPSTOSQ, node, stosDependencies, cg);
          break;
       default:
```

Comment:
Three of the case statements are identical except for the opcode value and the shift amount.  Consider refactoring to save on code bloat.